app/packages/looker/src/elements/common/util.ts

```diff
@@ -95,109 +94,150 @@ export const prettify = (
-          );
-        }
-        return l.value?.toString() == param[key]?.toString();
-      })?.color;
 
-      if (isValidColor(valueColor)) {
-        return valueColor;
-      }
+  /* In color by field mode, 
+  if ColorScheme setting has special color rule of current field,
+  if current field is a primitive field, find matching value in valueColors directly;
+  if current field is a embeddedDocumentField, use colorByAttribute as the key in param
+  to find matching value in valueColors. */
+
+  if (by === "value") {
+    if (!setting) {
+      return isPrimitive
+        ? fallbackColorByValuePrimitive
+        : getColor(pool, seed, fallbackLabel);
+    }
+
+    // For embeddedDocumentField, use colorByAttribute as the key in param or use
+    // the fallbackLabel defaulted to 'label', regression uses 'value'
+
+    const key = setting.colorByAttribute ?? fallbackLabel;
+    let currentValue;
+
+    if (!isPrimitive) {
+      const convertedKey = ![undefined, null].includes(param[key])
+        ? key
+        : fallbackLabel;
+      // use the first value as the fallback default if it's a listField for embeddedDoc
+      currentValue = !isPrimitive
+        ? Array.isArray(param[key])
+          ? param[convertedKey][0]
+          : param[convertedKey]
+        : null;
     } else {
-      key = labelDefault ? "label" : "value";
+      currentValue = value;
     }
-    return getColor(coloring.pool, coloring.seed, param[key]);
-  }
-  if (coloring.by === "value" && path === "tags") {
-    if (setting) {
+
+    // sample tags
+    if (path === "tags") {
       const valueColor = setting.valueColors?.find(
         (v) => v.value === param
       )?.color;
       if (isValidColor(valueColor)) {
         return valueColor;
+      } else {
+        return getColor(pool, seed, param);
       }
     }
-  }
-  return getColor(coloring.pool, coloring.seed, path);
-};
 
-// for primitive types
-type PrimitiveParam = {
-  coloring: Coloring;
-  path: string;
-  value: string | number | boolean | { datetime: number };
-  customizeColorSetting: CustomizeColor[];
-};
-export const getColorFromOptionsPrimitives = ({
-  coloring,
-  path,
-  value,
-  customizeColorSetting,
-}: PrimitiveParam) => {
-  const setting = customizeColorSetting.find((s) => s.path === path);
-  if (coloring.by === "field") {
-    if (isValidColor(setting?.fieldColor ?? "")) {
-      return setting?.fieldColor;
-    }
-    return getColor(coloring.pool, coloring.seed, path);
-  }
-  if (coloring.by === "value") {
-    if (setting) {
-      // check if this label has a assigned color, use it if it is a valid color
-      const valueColor = setting.valueColors?.find((l) => {
-        const normalized = l.value?.toString().toLowerCase();
-        if (["none", "null", "undefined"].includes(normalized)) {
+    // check if current attribute value or field value has an assigned color
+    const targetColor = setting.valueColors?.find((colorSetup) => {
```

Comment:
nit: If you could refactor this into its own utility and write some unit tests for it would be awesome